### PR TITLE
Case 21658: Push-To-Talk

### PR DESCRIPTION
**Pull request by r3tk0n**
Adds push-to-talk functionality to Interface. New setting will live under Settings > Audio for now, until the redesign of that menu is merged. The "T" key will be for Push to Talk (PTT). To use PTT in HMD, hold the grip buttons on both controllers.

This PR also divorces the Mute preference for HMD and Desktop modes, setting defaults to:
Desktop: Muted
HMD: Unmuted

User preferences will be respected between runs.

When the user enables, then disables PTT for a mode (Desktop / HMD), the mute setting for that mode will return to the default behavior. This was decided in a conversation between Wayne, Mukul, Joshua, and myself.

Manuscript ticket [here](https://highfidelity.manuscript.com/f/cases/21658/Push-To-Talk).

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17899/HighFidelity-Beta-PR15109-23e8a8b.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17899/HighFidelity-Beta-Interface-PR15109-23e8a8b.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17899/HighFidelity-Beta-PR15109-23e8a8b.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17899/HighFidelity-Quest-Beta-PR15109-23e8a8b.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17899/HighFidelity-Beta-PR15109-23e8a8b.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17899/HighFidelity-Beta-Interface-PR15109-23e8a8b.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17936/HighFidelity-Beta-PR15109-aabdb98.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17936/HighFidelity-Beta-Interface-PR15109-aabdb98.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17936/HighFidelity-Beta-PR15109-aabdb98.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17936/HighFidelity-Quest-Beta-PR15109-aabdb98.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17936/HighFidelity-Beta-PR15109-aabdb98.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17936/HighFidelity-Beta-Interface-PR15109-aabdb98.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17951/HighFidelity-Beta-PR15109-55efc31.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17951/HighFidelity-Quest-Beta-PR15109-55efc31.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17951/HighFidelity-Beta-PR15109-55efc31.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17951/HighFidelity-Beta-Interface-PR15109-55efc31.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17956/HighFidelity-Beta-PR15109-2b51f19.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/17956/HighFidelity-Beta-Interface-PR15109-2b51f19.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17956/HighFidelity-Beta-PR15109-2b51f19.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/17956/HighFidelity-Quest-Beta-PR15109-2b51f19.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17956/HighFidelity-Beta-PR15109-2b51f19.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/17956/HighFidelity-Beta-Interface-PR15109-2b51f19.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18029/HighFidelity-Beta-PR15109-04c6c42.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18029/HighFidelity-Beta-Interface-PR15109-04c6c42.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18029/HighFidelity-Beta-PR15109-04c6c42.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18029/HighFidelity-Beta-Interface-PR15109-04c6c42.exe

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18029/HighFidelity-Beta-PR15109-04c6c42.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18029/HighFidelity-Quest-Beta-PR15109-04c6c42.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18032/HighFidelity-Beta-PR15109-15eb90f.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18032/HighFidelity-Beta-Interface-PR15109-15eb90f.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18032/HighFidelity-Beta-PR15109-15eb90f.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18032/HighFidelity-Quest-Beta-PR15109-15eb90f.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18032/HighFidelity-Beta-PR15109-15eb90f.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18032/HighFidelity-Beta-Interface-PR15109-15eb90f.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18035/HighFidelity-Beta-PR15109-584342e.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18035/HighFidelity-Beta-Interface-PR15109-584342e.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18035/HighFidelity-Beta-PR15109-584342e.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18035/HighFidelity-Quest-Beta-PR15109-584342e.apk)

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18037/HighFidelity-Beta-PR15109-1b628bc.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18037/HighFidelity-Quest-Beta-PR15109-1b628bc.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18037/HighFidelity-Beta-PR15109-1b628bc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18037/HighFidelity-Beta-Interface-PR15109-1b628bc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18037/HighFidelity-Beta-PR15109-1b628bc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18037/HighFidelity-Beta-Interface-PR15109-1b628bc.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18039/HighFidelity-Beta-PR15109-c1ed011.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18039/HighFidelity-Beta-Interface-PR15109-c1ed011.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18039/HighFidelity-Beta-PR15109-c1ed011.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18039/HighFidelity-Quest-Beta-PR15109-c1ed011.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18039/HighFidelity-Beta-PR15109-c1ed011.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18039/HighFidelity-Beta-Interface-PR15109-c1ed011.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18043/HighFidelity-Beta-PR15109-8aedc98.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18043/HighFidelity-Beta-Interface-PR15109-8aedc98.dmg

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18043/HighFidelity-Beta-PR15109-8aedc98.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18043/HighFidelity-Quest-Beta-PR15109-8aedc98.apk)

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -4205,6 +4215,10 @@ void Application::keyPressEvent(QKeyEvent* event) {
                 }
                 break;
 
+            case Qt::Key_T:
+                audioScriptingInterface->setPushingToTalk(true);
```
Comment (ZappoMan) [inline interface/src/Application.cpp]: It would be preferable to do this using the keyboard.json input controller mapping...

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18043/HighFidelity-Beta-PR15109-8aedc98.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18043/HighFidelity-Beta-Interface-PR15109-8aedc98.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18054/HighFidelity-Beta-PR15109-4eefd13.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18054/HighFidelity-Beta-Interface-PR15109-4eefd13.dmg

File: interface/src/scripting/Audio.cpp [hunk 2]
```diff
@@ -40,6 +40,8 @@ Audio::Audio() : _devices(_contextIsHMD) {
     connect(client, &AudioClient::inputLoudnessChanged, this, &Audio::onInputLoudnessChanged);
     connect(client, &AudioClient::inputVolumeChanged, this, &Audio::setInputVolume);
     connect(this, &Audio::contextChanged, &_devices, &AudioDevices::onContextChanged);
+    // when pushing to talk changed, handle it.
```
Comment (danteruiz) [inline interface/src/scripting/Audio.cpp]: we do not need this comment.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18054/HighFidelity-Beta-PR15109-4eefd13.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18054/HighFidelity-Quest-Beta-PR15109-4eefd13.apk)

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18054/HighFidelity-Beta-PR15109-4eefd13.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18054/HighFidelity-Beta-Interface-PR15109-4eefd13.exe

File: interface/src/scripting/Audio.cpp (hunk 2 above)
Comment (r3tk0n) [reply interface/src/scripting/Audio.cpp]: Yuuuup, removed.

File: scripts/system/controllers/controllerModules/pushToTalk.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+Script.include("/~/system/libraries/controllerDispatcherUtils.js");
+Script.include("/~/system/libraries/controllers.js");
+
+(function () {		// BEGIN LOCAL_SCOPE
+    function PushToTalkHandler() {
+        var _this = this;
+        this.active = false;
+        //var pttMapping, mappingName;
+
+        this.setup = function () {
```
Comment (danteruiz) [inline scripts/system/controllers/controllerModules/pushToTalk.js]: what is this function used for?

Comment (wayne-chen) [reply scripts/system/controllers/controllerModules/pushToTalk.js]: fixed.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18066/HighFidelity-Beta-PR15109-db36833.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18066/HighFidelity-Quest-Beta-PR15109-db36833.apk)

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18067/HighFidelity-Beta-PR15109-0279373.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/18067/HighFidelity-Quest-Beta-PR15109-0279373.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18067/HighFidelity-Beta-PR15109-0279373.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/18067/HighFidelity-Beta-Interface-PR15109-0279373.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18067/HighFidelity-Beta-PR15109-0279373.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/18067/HighFidelity-Beta-Interface-PR15109-0279373.exe

**Pull request merged**